client/consul/consul.go

```diff
@@ -29,3 +36,101 @@ type ServiceIdentityAPI interface {
 type SupportedProxiesAPI interface {
 	Proxies() (map[string][]string, error)
 }
+
+// JWTLoginRequest is an object representing a login request with JWT
+type JWTLoginRequest struct {
+	JWT            string
+	Role           string
+	AuthMethodName string
+}
+
+// ConsulClient is the interface that the nomad client uses to interact with
+// Consul.
+type ConsulClient interface {
+	// DeriveSITokenWithJWT contacts the nomad server and fetches wrapped tokens
+	// for a set of tasks. The wrapped tokens will be unwrapped using consul and
+	// returned.
+	DeriveSITokenWithJWT(map[string]JWTLoginRequest) (map[string]string, error)
+}
+
+type consulClient struct {
+	// tokenDeriver is a function pointer passed in by the client to derive
+	// tokens by making RPC calls to the nomad server.
+	tokenDeriver TokenDeriverFunc
+
+	// client is the API client to interact with vault
+	client *consulapi.Client
+
+	// config is the configuration to connect to consul
+	config *config.ConsulConfig
+
+	logger hclog.Logger
+}
+
+// NewConsulClient creates a new Consul client
+func NewConsulClient(config *config.ConsulConfig, logger hclog.Logger, tokenDeriver TokenDeriverFunc) (*consulClient, error) {
+	if config == nil {
+		return nil, fmt.Errorf("nil consul config")
+	}
+
+	logger = logger.Named("consul")
+
+	c := &consulClient{
+		config:       config,
+		logger:       logger,
+		tokenDeriver: tokenDeriver,
+	}
+
+	// if UseIdentity is unset of set to false, return an empty client
+	if config.UseIdentity == nil || !*config.UseIdentity {
+		return nil, nil
+	}
```

Comment:
Returning `nil` seems to disagree with having the legacy token deriver as a field of the client struct. Should we `return c, nil` instead? If not, should we move this above everything else so we don't allocate a `consulClient` struct pointer?

Reply:
I moved it up in [4c925f3](https://github.com/hashicorp/nomad/pull/18370/commits/4c925f3a61b8b612482abe61fc7013dab06fd041),  I think there's no use in returning a client if the feature is disabled.